### PR TITLE
[HIVEMALL-145] Merge Brickhouse functions

### PR DESCRIPTION
**Pull request by myui**
## What changes were proposed in this pull request?

Merge [brickhouse](https://github.com/klout/brickhouse) functions.

## What type of PR is it?

Feature

## What is the Jira issue?

https://issues.apache.org/jira/browse/HIVEMALL-145

## How was this patch tested?

unit tests and manual tests

## How to use this feature?

as described in [user guide](http://hivemall.incubator.apache.org/userguide/misc/generic_funcs.html).

## Checklist

- [x] Did you apply source code formatter, i.e., `mvn formatter:format`, for your commit?
- [x] Did you run system tests on Hive (or Spark)?
- [x] Invite active/main Brickhouse developers as Hivemall PPMC members or committers.
https://github.com/klout/brickhouse/issues/149
- [x] +1 from Klout members to merge

### CODE REVIEW

**Comment by myui** [comment]
Still WIP for [reviewing](https://docs.google.com/spreadsheets/d/1gtFNcTvPR9OZAsbobj2D9d37tOx4nAoSlib9CLdEDQg/edit#gid=0) functions to merge.

**Comment by myui** [comment]
```sql
select 
  NAMED_STRUCT("Name", "John", "age", 31),
  to_json(
     NAMED_STRUCT("Name", "John", "age", 31)
  ),
  to_json(
     NAMED_STRUCT("Name", "John", "age", 31),
     array('Name', 'age')
  ),
  to_json(
     NAMED_STRUCT("Name", "John", "age", 31),
     array('name', 'age')
  ),
  to_json(
     NAMED_STRUCT("Name", "John", "age", 31),
     array('age')
  ),
  to_json(
     NAMED_STRUCT("Name", "John", "age", 31),
     array()
  ),
  to_json(
     null,
     array()
  ),
  to_json(
    struct("123", "456", 789, array(314,007)),
    array('ti','si','i','bi')
  ),
  to_json(
    struct("123", "456", 789, array(314,007)),
    'ti,si,i,bi'
  ),
  to_json(
    struct("123", "456", 789, array(314,007))
  ),
  to_json(
    NAMED_STRUCT("country", "japan", "city", "tokyo")
  ),
  to_json(
    NAMED_STRUCT("country", "japan", "city", "tokyo"), 
    array('city')
  ),
  to_json(
    ARRAY(
      NAMED_STRUCT("country", "japan", "city", "tokyo"), 
      NAMED_STRUCT("country", "japan", "city", "osaka")
    )
  ),
  to_json(
    ARRAY(
      NAMED_STRUCT("country", "japan", "city", "tokyo"), 
      NAMED_STRUCT("country", "japan", "city", "osaka")
    ),
    array('city')
  );
```

> {"name":"John","age":31}        {"name":"John","age":31}        {"Name":"John","age":31}        {"name":"John","age":31}        {"age":31}      {}NULL     {"ti":"123","si":"456","i":789,"bi":[314,7]}    {"ti":"123","si":"456","i":789,"bi":[314,7]}    {"col1":"123","col2":"456","col3":789,"col4":[314,7]}      {"country":"japan","city":"tokyo"}      {"city":"tokyo"}        [{"country":"japan","city":"tokyo"},{"country":"japan","city":"osaka"}]    [{"country":"japan","city":"tokyo"},{"country":"japan","city":"osaka"}]

```sql
select
  from_json(
    '{ "person" : { "name" : "makoto" , "age" : 37 } }',
    'struct<name:string,age:int>', 
    array('person')
  ),
  from_json(
    '[0.1,1.1,2.2]',
    'array<double>'
  ),
  from_json(to_json(
    ARRAY(
      NAMED_STRUCT("country", "japan", "city", "tokyo"), 
      NAMED_STRUCT("country", "japan", "city", "osaka")
    )
  ),'array<struct<country:string,city:string>>'),
  from_json(to_json(
    ARRAY(
      NAMED_STRUCT("country", "japan", "city", "tokyo"), 
      NAMED_STRUCT("country", "japan", "city", "osaka")
    ),
    array('city')
  ), 'array<struct<country:string,city:string>>'),
  from_json(to_json(
    ARRAY(
      NAMED_STRUCT("country", "japan", "city", "tokyo"), 
      NAMED_STRUCT("country", "japan", "city", "osaka")
    )
  ),'array<struct<city:string>>');
```

> {"name":"makoto","age":37}      [0.1,1.1,2.2]   [{"country":"japan","city":"tokyo"},{"country":"japan","city":"osaka"}] [{"country":"japan","city":"tokyo"},{"country":"japan","city":"osaka"}]    [{"city":"tokyo"},{"city":"osaka"}]

**Comment by myui** [comment]
@jeromebanks merging of Brickhouse functions is in-progress in this PR. FYI

We need to add unit test, improve qualities of functions, and add documents.

**Comment by myui** [comment]
```sql
create temporary function moving_avg as 'hivemall.statistics.MovingAverageUDTF';

select moving_avg(x, 3) from (select explode(array(1,2,3,4,5,6,7)) as x) series;
select moving_avg(x, 3) from (select explode(array(1.0,2.0,3.0,4.0,5.0,6.0,7.0)) as x) series;
```

|avg|
|:-:|
|1.0|
|1.5|
|2.0|
|3.0|
|4.0|
|5.0|
|6.0|

**Comment by myui** [comment]
@maropu Could you check whether `to_json` and `from_json` works on Spark or not if possible?

I'm not sure hcatalog is provided in Spark environment. 
https://github.com/apache/incubator-hivemall/pull/135/commits/dd99307ef49f0507870573efdf5f2ae8da240cca#diff-357e4854869b2e21c38b1b437f11095aR56

**Comment by myui** [comment]
```sql
create temporary function conditional_emit as 'hivemall.tools.array.ConditionalEmitUDTF';

WITH input as (
   select array(true, false, true) as conditions, array("one", "two", "three") as features
   UNION ALL
   select array(true, true, false), array("four", "five", "six")
)
select
  conditional_emit(
     conditions, features
  )
from 
  input;
```

|feature|
|:-:|
|one|
|three|
|four|
|five|

**Comment by maropu** [comment]
@myui Spark already has these functions: https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/functions.scala#L3118

**Comment by myui** [comment]
@maropu Deprecated SubarrayUDF to use ArraySliceUDF instead. FYI

**Comment by paulojblack** [comment]
Just a heads up, I think some of the changes here that were pushed to master recently have broken set-up as instructed in the getting started docs. 

Specifically having trouble with the changes in https://github.com/apache/incubator-hivemall/blob/master/resources/ddl/define-all.hive. After commenting out lines 409-413 it works as expected.

**Comment by myui** [comment]
@paulojblack Thank you for comments. Will confirm it and fix master.

**Comment by myui** [comment]
@jeromebanks Add you to [the committer list](http://hivemall.incubator.apache.org/team-list.html) with https://github.com/apache/incubator-hivemall/pull/135/commits/e956e98fdc80f3498cad8e55bc88cac1518c5f30
FYI

**Comment by myui** [comment]
@paulojblack You need to use up-to-date DDLs since we updated DDLs for `subarray` UDF in https://github.com/apache/incubator-hivemall/pull/135/commits/7003006e1c27cae66d5aa5c91fccf27a21b105a7

By using [define-all.hive](https://github.com/apache/incubator-hivemall/blob/master/resources/ddl/define-all.hive) in master branch, it's working without errors in my environment.

**Comment by myui** [comment]
If you are using `v0.5.0`, then you need to use [one of v0.5.0](https://github.com/apache/incubator-hivemall/blob/v0.5.0/resources/ddl/define-all.hive).

DDLs are pointing specified release branches in [the distribution page](http://hivemall.incubator.apache.org/download.html).

[Installation manual](http://hivemall.incubator.apache.org/userguide/getting_started/installation.html) can be improved though.

**Comment by paulojblack** [comment]
That makes sense, I figured a change like that wasnt made blindly. Consider it a heads up on the docs then!

**Comment by myui** [comment]
@paulojblack Generally, we recommend to use [Official ASF releases](http://hivemall.incubator.apache.org/download.html), not one in the master branch.

When you are using the master branch, use the latest DDLs with a caution.

**Comment by myui** [comment]
```sql
select generate_series(2,4);

value
2
3
4

select generate_series(5,1,-2);

value
5
3
1

select generate_series(4,3);

(no return)

select date_add(current_date(),value) as `date`,value from (select generate_series(1,3)) t;

date    value
2018-04-21      1
2018-04-22      2
2018-04-23      3

WITH input as (
 select 1 as c1, 10 as c2, 3 as step
 UNION ALL
 select 10, 2, -3
)
select generate_series(c1, c2, step) as series from input;

series
1
4
7
10
10
7
4
```

**Comment by myui** [comment]
```sql
create temporary function merge_maps as 'hivemall.tools.map.MergeMapsUDAF';

create table test as 
 SELECT map('A',10,'B',20,'C',30) as m
 UNION ALL 
 SELECT map('A',11,'D',40,'E',50) as m;

> {"A":11,"B":20,"C":30,"D":40,"E":50}

SELECT merge_maps(m) FROM test;
```

File: core/src/main/java/hivemall/tools/sanity/AssertUDF.java [hunk 1]
```diff
@@ -25,8 +25,10 @@
 
 @Description(name = "assert",
         value = "_FUNC_(boolean condition) or _FUNC_(boolean condition, string errMsg)"
-                + "- Throws HiveException if condition is not met")
-@UDFType(deterministic = true, stateful = false)
+                + "- Throws HiveException if condition is not met",
+        extended = "SELECT count(1) FROM stock_price WHERE assert(price > 0.0);\n"
+                + "SELECT count(1) FROM stock_price WHRE assert(price > 0.0, 'price MUST be more than 0.0')")
```
Comment (takuti) [inline core/src/main/java/hivemall/tools/sanity/AssertUDF.java]: typo `s/WHRE/WHERE/`

**Comment by jeromebanks** [comment]
@myui - Sure.  I've been silently lurking, not sure when to step in. Looks fine in general, +1.  I will do more in depth review however.
--- @jeromebanks
    On Tuesday, June 5, 2018, 3:23:58 AM PDT, Makoto YUI <notifications@github.com> wrote:  
 
 
@jeromebanks I'm considering to merge this PR. Could you review if possible?

**Comment by myui** [comment]
For K-minimum Values (KMV) and Sketch related codes, I'll create an another JIRA ticket.

For other UDFs, we accept incoming PRs.
https://docs.google.com/spreadsheets/d/1gtFNcTvPR9OZAsbobj2D9d37tOx4nAoSlib9CLdEDQg/edit#gid=0

**Pull request closed without merging**